Made About HTML "Internationalizable"

virtuoushub opened the pull request:
After reading @i-make-robots' [comment](https://github.com/MarginallyClever/Makelangelo/pull/61#issuecomment-85517620), I pulled out the hardcoded About Message Dialog's HTML String constant into the translation file.

Because of the dynamic way the HTML is built in the Java code, I had to split the internationalized HTML String in two. I added the key/values to the all the languages, FIXMEs where it is not translated, and documentation as to why the key/values are split up in the Java and XML files.

i-make-robots commented:
i would have used a unique token like CURRENT_VERSION and then run a string replace.  6 of one, half a dozen of the other.  Nice work!

virtuoushub commented:
@i-make-robots, thanks. I was hoping for a cleaner solution. The CURRENT_VERSION approach you mentioned was something I thought about but reminded me too much of a [magic number](http://en.wikipedia.org/wiki/Magic_number_%28programming%29), so I did not go forward with it. With proper documentation I do not think it would have [smelled](http://en.wikipedia.org/wiki/Code_smell) any more than my approach.

I was not sure how to cleanly solve this short of using something like XSLT which, for the problem domain, seems like overkill.

I tried to abstract the HTML string building away so that in the future if a better approach is found, we just swap that implementation in.

i-make-robots commented:
All sensible considerations.  Clearly you know what you're doing.
Any plans to tinker elsewhere?  I'm planning to switch out the GFX for
opengl, should greatly improve the render time and the UI responsiveness.

On Wed, Mar 25, 2015 at 7:47 AM, Peter Colapietro <notifications@github.com>
wrote:

> @i-make-robots <https://github.com/i-make-robots>, thanks. I was hoping
> for a cleaner solution. The CURRENT_VERSION approach you mentioned was
> something I thought about but reminded me too much of a magic number
> <http://en.wikipedia.org/wiki/Magic_number_%28programming%29>, so I did
> not go forward with it. With proper documentation I do not think it would
> have smelled <http://en.wikipedia.org/wiki/Code_smell> any more than my
> approach.
>
> I was not sure how to cleanly solve this short of using something like
> XSLT which, for the problem domain, seems like overkill.
>
> I tried to abstract the HTML string building away so that in the future if
> a better approach is found, we just swap that implementation in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarginallyClever/Makelangelo/pull/63#issuecomment-86062810>
> .
>



-- 
Dan Royer :: Marginally Clever <http://www.marginallyclevergroup.com/> ::
Raising Robot Literacy :: +1.604.916.2281

virtuoushub commented:
@i-make-robots, yes my plan is to iteratively add in features as I look more into the code base and [GitHub issues](https://github.com/MarginallyClever/Makelangelo/issues). Is there something specific you would like me to take a look at?

tl;dr I also want to add a build system, and improve the version naming scheme.

I was also thinking about how to make the build more "developer environment agnostic." In the [branch I am working out of](https://github.com/MakeHartford/Makelangelo/tree/dev), I added in [Maven](http://maven.apache.org/) support, although I do have a slight preference for [Gradle](http://www.gradle.org/) so I might switch over to that. I saw what seemed to be an auto generated Ant [build file](https://github.com/MarginallyClever/Makelangelo/blob/master/antscript.xml) in the code, so if that is your preference I would be happy to incorporate it into my flow to ensure proper building across developers' environments. The beauty of a lot of these build systems is that it is relatively painless* to get them to play nice with each other if you are so inclined to wire them up to do so.

Since it is your project, I wanted to make any transition to a new build system as transparent and painless as possible so that you can keep working with your current flow and not have it be disrupted. 

My thinking is that once a build system is in place, the project will be more aptly suited for continuous [integration](http://en.wikipedia.org/wiki/Continuous_integration) and [delivery](http://en.wikipedia.org/wiki/Continuous_delivery).

Might I also recommend possibly transitioning to a version naming scheme such as [Semantic Versioning](http://semver.org)? In my fork, I added in a version tag for [v6](https://github.com/MakeHartford/Makelangelo/tree/v6). To transition to Semantic Versioning for the next version would be as simple as tagging that release as `v7.0.0`. If you choose to adopt Semantic Versioning `v6.1.0` might also suffice** depending on what was added in-between releases.  In conjunction with a build system I find that maintenance and delivery of builds can be drastically improved using this roughly outlined approached.

___*___ Do not quote me on this even though I just publicly stated it. :eyeglasses::dog:
___**___ Forgive me as I don't know the full semver specifiction. :sunglasses:

i-make-robots commented:
I'm OK with all these changes.  I'd love to get to the point where the app
is a native-branded file that's click-to-run, instead of batch-file-patch
BS we have now.  I'd also really like to see "download and update now?
Y/N"  My Java skills aren't that good, and your suggestions are all
approved.  :)

Do people need to know the version number if we have continuous integration
and delivery?  All they care is that something's been made better enough
that it's worth downloading.



On Wed, Mar 25, 2015 at 8:48 AM, Peter Colapietro <notifications@github.com>
wrote:

> @i-make-robots <https://github.com/i-make-robots>, yes my plan is to
> iteratively add in features as I look more into the code base and GitHub
> issues <https://github.com/MarginallyClever/Makelangelo/issues>. Is there
> something specific you would like me to take a look at?
>
> tl;dr I also want to add a build system, and improve the version naming
> scheme.
>
> I was also thinking about how to make the build more "developer
> environment agnostic." In the branch I am working out of
> <https://github.com/MakeHartford/Makelangelo/tree/dev>, I added in Maven
> <http://maven.apache.org/> support, although I do have a slight
> preference for Gradle <http://www.gradle.org/> so I might switch over to
> that. I saw what seemed to be an auto generated Ant build file
> <https://github.com/MarginallyClever/Makelangelo/blob/master/antscript.xml>
> in the code, so if that is your preference I would be happy to incorporate
> it into my flow to ensure proper building across developers' environments.
> The beauty of a lot of these build systems is that it is relatively
> painless* to get them to play nice with each other if you are so inclined
> to wire them up to do so.
>
> Since it is your project, I wanted to make any transition to a new build
> system as transparent and painless as possible so that you can keep working
> with your current flow and not have it be disrupted.
>
> My thinking is that once a build system is in place, the project will be
> more aptly suited for continuous integration
> <http://en.wikipedia.org/wiki/Continuous_integration> and delivery
> <http://en.wikipedia.org/wiki/Continuous_delivery>.
>
> Might I also recommend possibly transitioning to a version naming scheme
> such as Semantic Versioning <http://semver.org>? In my fork, I added in a
> version tag for v6 <https://github.com/MakeHartford/Makelangelo/tree/v6>.
> To transition to Semantic Versioning for the next version would be as
> simple as tagging that release as v7.0.0. If you choose to adopt Semantic
> Versioning v6.1.0 might also suffice** depending on what was added
> in-between releases. In conjunction with a build system I find that
> maintenance and delivery of builds can be drastically improved using this
> roughly outlined approached.
>
> *** Do not quote me on this even though I just publicly stated it. [image:
> :eyeglasses:][image: :dog:]
> **** Forgive me as I don't know the full semver specifiction. [image:
> :sunglasses:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MarginallyClever/Makelangelo/pull/63#issuecomment-86090414>
> .
>



-- 
Dan Royer :: Marginally Clever <http://www.marginallyclevergroup.com/> ::
Raising Robot Literacy :: +1.604.916.2281

virtuoushub commented:
@i-make-robots, no, the end user does not need to know the version number, it is more for the developer. However, if there are issues with the build, having the exact version number in a bug report allows for much quicker debugging/issue resolving.

To your point about auto update, I agree. Leapfrogging off of what I was just talking about, say the app crashes having a "submit crash log? Y/N" is something I always wanted to implement. I will first look into auto updating, as that is also something I always wanted to implement and is of higher priority by my calculations.
